cmp/options.go

```diff
@@ -183,9 +183,17 @@ func Ignore() Option {
 // The transformer f must be a function "func(T) R" that converts values of
 // type T to those of type R and is implicitly filtered to input values
 // assignable to T. The transformer must not mutate T in any way.
-// If T and R are the same type, an additional filter must be applied to
-// act as the base case to prevent an infinite recursion applying the same
-// transform to itself (see the SortedSlice example).
+//
+// To help prevent infinite cycles recursively applying the same transform to
+// the output of itself (e.g., in the case where the input and output types
+// are the same), an implicit filter is added such that a transformer is
+// applicable only if that exact transformer is not already in the
+// tail of the Path since the last non-Transform step.
+// While this is sufficient for most usages of Transformers, this does not
```

Comment:
Both.

I haven't looked at your actual code, but my thinking is that once you apply a transformer to a value in a given invocation of cmp.Equal, you shouldn't apply it again, whether or not it changes the path.

Basically, I'm thinking in terms of invocations of cmp.Equal, not path changes. Every invocation of cmp.Equal changes the path, but not every path change is the result of a call on cmp.Equal. I didn't realize until this discussion that you put things like type assertions and transforms in the path, and I don't totally get why you do, but in any case I'd ignore all the path steps that don't correlate with a call on Equal.